File: ocaml/lib/day2.ml
type color =
  | Blue of int
  | Red of int
  | Green of int

type game_totals =
  { blue : int list
  ; red : int list
  ; green : int list
  }

let init_color string num =
  match string with
  | "blue" -> Blue num
  | "red" -> Red num
  | "green" -> Green num
  | _ -> failwith "invalid color"
;;

let init_game_total () = { blue = []; red = []; green = [] }

let add_color color game_total =
  match color with
  | Blue num -> { game_total with blue = num :: game_total.blue }
  | Red num -> { game_total with red = num :: game_total.red }
  | Green num -> { game_total with green = num :: game_total.green }
;;

let find_max lst =
  match lst with
  | [] -> failwith "empty list"
  | hd :: tl -> List.fold_left max hd tl
;;

let get_total game_total =
  let blue_amt = find_max game_total.blue in
  let red_amt = find_max game_total.red in
  let green_amt = find_max game_total.green in
  blue_amt * red_amt * green_amt
;;

let print_color color =
  match color with
  | Green num -> Printf.printf "green %d\n" num
  | Red num -> Printf.printf "red %d\n" num
  | Blue num -> Printf.printf "blue %d\n" num
;;

let is_possible color =
  match color with
  | Green num when num <= 13 -> true
  | Red num when num <= 12 -> true
  | Blue num when num <= 14 -> true
  | _ -> false
;;

let p1 input =
  let res =
    String.split_on_char '\n' input
    |> List.filter (fun line ->
      match line with
      | "" -> false
      | _ -> true)
    |> List.fold_left
         (fun acc line ->
           let split = String.split_on_char ':' line in
           let game_id =
             List.nth (String.split_on_char ' ' (List.nth split 0)) 1
             |> int_of_string
           in
           let set_strs = List.nth split 1 |> String.split_on_char ';' in
           let games =
             List.fold_left
               (fun acc set ->
                 let game = String.split_on_char ',' set in
                 let games =
                   List.fold_left
                     (fun acc g ->
                       let trimmed = String.trim g in
                       let split_trimmed = String.split_on_char ' ' trimmed in
                       let num = List.nth split_trimmed 0 |> int_of_string in
                       let color_str = List.nth split_trimmed 1 in
                       let color = init_color color_str num in
                       color :: acc)
                     []
                     game
                 in
                 games :: acc)
               []
               set_strs
           in
           let flat = List.flatten games in
           let possible =
             List.fold_left
               (fun acc cur -> if is_possible cur then acc else false)
               true
               flat
           in
           if possible then acc + game_id else acc)
         0
  in
  res
;;

let p2 input =
  let res =
    String.split_on_char '\n' input
    |> List.filter (fun line ->
      match line with
      | "" -> false
      | _ -> true)
    |> List.fold_left
         (fun acc line ->
           let split = String.split_on_char ':' line in
           let set_strs = List.nth split 1 |> String.split_on_char ';' in
           let games =
             List.fold_left
               (fun acc set ->
                 let game = String.split_on_char ',' set in
                 let games =
                   List.fold_left
                     (fun acc g ->
                       let trimmed = String.trim g in
                       let split_trimmed = String.split_on_char ' ' trimmed in
                       let num = List.nth split_trimmed 0 |> int_of_string in
                       let color_str = List.nth split_trimmed 1 in
                       let color = init_color color_str num in
                       color :: acc)
                     []
                     game
                 in
                 games :: acc)
               []
               set_strs
           in
           let totals_record =
             List.flatten games
             |> List.fold_left (fun acc g -> add_color g acc)
                @@ init_game_total ()
           in
           acc + get_total totals_record)
         0
  in
  res
;;

File: cpp/src/day2.cc
#include "aoc.hh"
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

int day2p1(const std::string& input) {
    std::istringstream iss(input);
    std::string line;
    int res = 0;

    while (std::getline(iss, line, '\n')) {
        std::vector<std::string> split = split_str(line, ':');
        std::string& game_id_str = split[0];
        std::string& game_str = split[1];
        std::string& id_str = game_id_str.replace(0, 5, "");
        int id = std::stoi(id_str);

        std::vector<std::string> pulls = split_str(game_str, ';');
        bool possible = true;
        for (auto& pull_str : pulls) {
            std::string pull = trim(pull_str);
            std::vector<std::string> cube_strs = split_str(pull, ',');
            for (auto& cube_str_ : cube_strs) {
                std::string cube_str = trim(cube_str_);
                std::vector<std::string> num_and_color =
                    split_str(cube_str, ' ');
                int number = std::stoi(num_and_color[0]);
                std::string& color = num_and_color[1];
                if (color == "green") {
                    if (number > 13) {
                        possible = false;
                    }
                } else if (color == "blue") {
                    if (number > 14) {
                        possible = false;
                    }
                } else if (color == "red") {
                    if (number > 12) {
                        possible = false;
                    }
                }
            }
        }
        if (possible) {
            res += id;
        }
    }
    return res;
}

int day2p2(const std::string& input) {
    std::istringstream iss(input);
    std::string line;
    int res = 0;

    while (std::getline(iss, line, '\n')) {
        std::vector<std::string> split = split_str(line, ':');
        std::string& game_str = split[1];

        std::vector<std::string> pulls = split_str(game_str, ';');
        int full;
        int green_amt = 0;
        int blue_amt = 0;
        int red_amt = 0;
        for (auto& pull_str : pulls) {
            std::string pull = trim(pull_str);
            std::vector<std::string> cube_strs = split_str(pull, ',');
            for (auto& cube_str_ : cube_strs) {
                std::string cube_str = trim(cube_str_);
                std::vector<std::string> num_and_color =
                    split_str(cube_str, ' ');
                int number = std::stoi(num_and_color[0]);
                std::string& color = num_and_color[1];
                if (color == "green") {
                    if (number > green_amt) {
                        green_amt = number;
                    }
                } else if (color == "blue") {
                    if (number > blue_amt) {
                        blue_amt = number;
                    }
                } else if (color == "red") {
                    if (number > red_amt) {
                        red_amt = number;
                    }
                }
            }
        }
        full = blue_amt * red_amt * green_amt;
        res += full;
    }
    return res;
}
